src/dynamic_boundary_conditions/tide_data.py

```diff
@@ -1,148 +1,458 @@
-specify how many days of data is required and how much data (in days) either side of the highest tide value they
-want. They are then outputted a dataframe selection
-that includes the highest tide and either side data.
-"""
-
-# Imported packages
-import os
-from dotenv import load_dotenv
+import logging
+import pathlib
 from datetime import date, datetime, timedelta
+from typing import Dict, List, Tuple, Union, Optional
+
 import pandas as pd
-from shapely.geometry import Polygon
-import aiohttp
+import geopandas as gpd
 import asyncio
-from src.digitaltwin import setup_environment
-
-
-def get_all_date(start_date: str, total_days: int, max_range: int = 30) -> dict:
-    """ get each start date and number of days for total days. """
-    start_date = datetime.strptime(start_date, "%Y-%m-%d").date()
-    start_date_dict = {start_date: 0}
-    while total_days > 0:
-        if total_days > max_range:
-            start_date_dict[(sorted(start_date_dict.keys()))[-1]] = 30
-            start_date_dict[(sorted(start_date_dict.keys())[-1] + timedelta(days=max_range))] = total_days - 31
-        else:
-            start_date_dict[(sorted(start_date_dict.keys()))[-1]] = total_days
-        total_days -= 30
-    return start_date_dict
-
-
-def gen_param_list(start_date_dict: dict, request_parameters: dict) -> list:
-    """ generate a parameters list for all request. """
-    params_list = []
-    for start_date, number_of_days in start_date_dict.items():
-        params = request_parameters.copy()
-        params['startDate'] = start_date.isoformat()
-        params['numberOfDays'] = number_of_days
-        params_list.append(params)
-    return params_list
-
-
-async def get(session: aiohttp.ClientSession, url: str, params: dict) -> dict:
-    """ request get data from url """
-    resp = await session.request('GET', url=url, params=params)
-    return await resp.json()
-
-
-async def get_tide_dataframe(params_list: list, url: str = 'https://api.niwa.co.nz/tides/data') -> pd.DataFrame:
-    """ Asynchronous context manager. """
+import aiohttp
+
+from src import config
+from src.dynamic_boundary_conditions.tide_enum import DatumType
+
+log = logging.getLogger(__name__)
+log.setLevel(logging.DEBUG)
+
+formatter = logging.Formatter("%(levelname)s:%(asctime)s:%(name)s:%(message)s")
+stream_handler = logging.StreamHandler()
+stream_handler.setFormatter(formatter)
+
+log.addHandler(stream_handler)
+
+
+def get_catchment_area_coords(catchment_file: pathlib.Path) -> Tuple[float, float]:
+    """
+    Extract the catchment polygon centroid coordinates.
+
+    Parameters
+    ----------
+    catchment_file : pathlib.Path
+        The file path for the catchment polygon.
+    """
+    catchment = gpd.read_file(catchment_file)
+    catchment = catchment.to_crs(4326)
+    catchment_polygon = catchment["geometry"][0]
+    long, lat = catchment_polygon.centroid.coords[0]
+    return lat, long
+
+
+def get_date_ranges(
+        start_date: str,
```

Comment:
I think `start_date` should be type `datetime.date` so that we don't have to do user validation so low down

Reply:
and then this function would return `Dict[datetime.date, int]`

Reply:
or equivalently `List[Tuple[datetime.date, int]]`

Reply:
I've made some changes. Please review it again.